### PR TITLE
New actor math.

### PR DESCRIPTION
**Pull request by martun**

### CODE REVIEW

File: include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp [hunk 1]
```diff
@@ -51,10 +54,14 @@ namespace nil {
...
                     typedef typename FieldType::value_type value_type;
                     cache.resize(size);
-                    cache[0] = value_type::one();
-                    for (std::size_t i = 1; i < size; ++i) {
-                        cache[i] = cache[i - 1] * omega;
-                    }
+                    wait_for_all(ThreadPool::get_instance(ThreadPool::PoolLevel::LOW).block_execution<void>(
+                        size,
+                        [&cache, &omega](std::size_t begin, std::size_t end) {
+                            cache[begin] = omega.pow(begin);
```
Comment (Zerg1996) [inline include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp:60]: Please remind me - omega.pow - O(N) or it's smth like O(log N)

File: include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp [hunk 2]
```diff
@@ -74,27 +81,52 @@ namespace nil {
...
+                            if (k < rk)
+                                std::swap(a[k], a[rk]);
+                        }
+                    );
+
 
                     // invariant: m = 2^{s-1}
                     value_type t;
-                    for (std::size_t s = 1, m = 1, inc = n / 2; s <= logn; ++s, m <<= 1, inc >>= 1) {
+                    for (std::size_t s = 1, m = 1, inc = n / 2; s <= logn; ++s, m <<= 1, inc >>=1) {
```
Comment (Zerg1996) [inline include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp]: Let's not remove whitespace

File: include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp (hunk 1 above)
Comment (martun) [reply include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp:60]: It's O(log(N)), and in practice is very fast.

File: include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp (hunk 2 above)
Comment (martun) [reply include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp]: Added back.

File: include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp [hunk 3]
```diff
@@ -108,8 +140,13 @@ namespace nil {
                     if (omega_cache == nullptr) {
                         std::vector<typename FieldType::value_type> omega_powers;
                         create_fft_cache<FieldType>(a.size(), omega, omega_powers);
+if (omega_powers.size() == 0)
+    throw "We have a problem here.";
```
Comment (Zerg1996) [inline include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp]: If we use throw - when we need to implement catch somewhere... or just use assert.

File: test/polynomial_dfs.cpp [hunk 4]
```diff
@@ -1294,6 +1296,40 @@ BOOST_AUTO_TEST_CASE(polynomial_dfs_zero_one_test) {
...
+
+    nil::crypto3::parallel_for(
+        0, poly4.size(),
+        [&poly4, &poly](std::size_t i) {
+            // Inside this multiplication lower level pool is used.
+            poly4[i] *= poly;
+        },
+        nil::crypto3::ThreadPool::PoolLevel::HIGH);
+
+    for (int i = 1; i < poly4.size(); ++i) {
```
Comment (Zerg1996) [inline test/polynomial_dfs.cpp]: Mb size_t for i

File: include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp (hunk 3 above)
Comment (martun) [reply include/nil/crypto3/math/domains/detail/basic_radix2_domain_aux.hpp]: Thanks, this was a leftover from my debugging section, it does not belong here. Removed.

File: test/polynomial_dfs.cpp (hunk 4 above)
Comment (martun) [reply test/polynomial_dfs.cpp]: Changed.

**Review by Zerg1996 — APPROVED** [review]
lgtm!

**Pull request merged**